miri

```diff
@@ -156,6 +159,10 @@ run|run-debug)
     # Then run the actual command.
     exec cargo run $CARGO_BUILD_FLAGS -- --sysroot "$MIRI_SYSROOT" "$@"
     ;;
+fmt)
+    find "$MIRIDIR" -not \( -name target -prune \) -name '*.rs' \
+        | xargs rustfmt --edition=2021 --config-path "$MIRIDIR/rustfmt.toml" "$@"
```

Comment:
This will pick up whichever rustfmt is in the override, so it would still try to use the non nightly one

Reply:
I intended that. Based on https://github.com/rust-lang/miri/pull/2257#discussion_r903995863 we (at least Ralf) are not expecting divergence in formatting between nightly and the miri toolchain. `rust-version` is updated often enough (multiple times per week, judging by https://github.com/rust-lang/miri/commits/master/rust-version) that there can't really be prolonged disagreement in formatting between newest nightly and current miri rust-version.

Using rustfmt from the miri toolchain here eliminates a stumbling block compared to `rustfmt +nightly` of the user not having a toolchain called `nightly` installed, or not having the `rustfmt` component installed for their nightly toolchain, which are both common scenarios.

Reply:
That's a weird mismatch between CI and local use though.

My intention all along was that we should do one of the following:
- Always use nightly for formatting, both on CI and in the usual dev workflow.
- Always use the `rust-version` toolchain for formatting, both on CI and in the usual dev workflow.

My personal preference is for the former option. It is also the smaller change, so it would make sense to start with that for `./miri fmt`. Then we can separately discuss if we want to move formatting to the `rust-version` toolchain (which on CI would likely mean to merge the formatting job with the `clippy` job so we avoid installing the same toolchain too often).

Reply:
> That's a weird mismatch between CI and local use though.

You can't have it both ways&mdash;

- **either** rustfmt's formatting does *not* change often enough to cause friction, in which case using nightly rustfmt in CI (for fast install) and miri rust-version in local workflow (for consistent+reliable developer experience) is fine;

- **or** rustfmt's formatting does tend to churn, in which case anything involving "most recent nightly's" unpinned rustfmt is a bad solution. Developers will not always have the same one installed. Developers having the same one installed is the point of `rust-version`.

> My personal preference is for the former option. It is also the smaller change, so it would make sense to start with that for `./miri fmt`.

I recognize that this is not my call, but I do not agree with this. I would dislike formatting being enforced in CI (even if ignored by bors) if it's going to cause red signal on PRs due to what nightly was on the contributor's system (I categorize this as "not the contributor's fault"). As a contributor I would prefer either no red signal related to formatting, or formatting that is tied to a specific pinned version of the tooling, just like the rest of the project. Formatting change, on the rare occasion they happen, can be landed atomically with the `rust-version` bump.

My favorite implementation is the one before the most recent force-push (145fac78cbb83a437e259b046cdab3a135325307).

- Rustfmt version in local workflow and CI are both governed by `rust-version`, same as the rustc version that contributors already need to keep up with.
- Faster than the current fmt job in GitHub Actions, even while checking 10&times; as many files.
- Does not block clippy from running or vice versa. You get signal about both on every run.

The downside identified above is that it duplicated 97 bytes from `./miri` into `ci.yml` but I don't personally see a problem with this, especially against the advantages over the alternatives.

Reply:
> You can't have it both ways—

My concern is not that this causes formatting mismatches, but just that it is inconsistent.

> I recognize that this is not my call, but I do not agree with this. I would dislike formatting being enforced in CI (even if ignored by bors) if it's going to cause red signal on PRs due to what nightly was on the contributor's system (I categorize this as "not the contributor's fault"). As a contributor I would prefer either no red signal related to formatting, or formatting that is tied to a specific pinned version of the tooling, just like the rest of the project. Formatting change, on the rare occasion they happen, can be landed atomically with the rust-version bump.

Yes, that is a good argument for using the rustfmt from `rust-version`. I have not experienced any trouble due to our lack of rustfmt pinning yet, but that might be pure luck. I am, usually, quite in favor of pinning everything to avoid hard-to-control variables.

I was just hoping to separate the addition of `./miri fmt` from the discussion of whether we want to move to a pinned rustfmt.

> Faster than the current fmt job in GitHub Actions, even while checking 10× as many files.

That `dtolnay@install` job is quite neat. :D  It is yet another new dependency though, and if we use it here we should probably use it at least also for the clippy jobs (the main build jobs already use caching so they should be fine).

> Does not block clippy from running or vice versa. You get signal about both on every run.

That is nice-to-have but not critical IMO. I'd rather avoid duplicating the RTIM invocation even if it means we lose a bit of signal fidelity here. Of course ideally GHA would provide a way to share some "setup" code between jobs so we could do this with less duplication in our CI config, but that is another topic entirely...

---

So, it looks like we are not going to get this PR landed without making a decision about pinned vs not-pinned rustfmt. (If we go for not-pinned rustfmt I don't think we should have `-c rustfmt` in `./rustup-toolchain`.) @oli-obk what do you think? At this point I feel I am sufficiently alone with my stance on rustfmt that I am willing to give in to the majority and make it mandatory and run it with the clippy job: https://github.com/rust-lang/miri/pull/2261.

Reply:
I do prefer pinning and enforcing formatting in CI, but imo that's besides the point. Miri is not a democracy and you're the main maintainer and contributor. Whatever works best for your workflow is fine with me and I adjust. Right now I'm running `cargo +nightly fmt && ./miri bless`, just like in rustc I run `./x.py fmt && ./x.py whatever`. Works for me

Reply:
:+1: Will rebase on #2261 and use `rustfmt +nightly` in the implementation of `./miri fmt`

Reply:
We could automatically download clippy/rustfmt on first use, that would avoid downloading it for ppl who just need a locally compiled miri without contributing code

Reply:
How would you detect first use?

I think for now we can add it to `rustup-toolchain`, since as @dtolnay said it's just a 2-3% increase of download size (adding both of them). If I end up feeling like this is a noticeable slowdown (which seems unlikely), we can figure something out.

Reply:
> How would you detect first use?

By trying to run `rustfmt --version` and installing the component if that fails